book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -165,49 +220,232 @@ each rooted at the highest finalized block. Each chain consists of a map from
+
+1. Update cummulative data members
+    - Remove the corresponding hash from `self.height_by_hash`
+    - for each `transaction` in `block`
+      - remove `transaction.hash` from `tx_by_hash`
+    - Add consumed utxos and remove new utxos from `self.utxos`
+    - Remove anchors from the appropriate `self.<version>_anchors`
+    - Remove the nullifiers from the appropriate `self.<version>_nullifiers`
+    - Subtract work from `self.partial_cumulative_work`
+
+1. Return the block
 
-Otherwise, attempt to perform contextual validation checks and the commit the
-given block to the state. The exact list of contextual validation checks will
-be specified in a later RFC. If contextual validation checks succeed, commit
-the new blocks to the non-finalized state as described below. Next, if the
-resulting non-finalized chain is longer than 100 blocks, the oldest block is
-now past the reorg limit. Remove it from the non-finalized state and commit
-it as a finalized block, as described in the next section.
+#### `Ord`
 
-XXX: fill in details of non-finalized state.
+The `Chain` type also implements `Ord` for reorganizing chains. First chains
+are compared by their `partial_cummulative_work`. Ties are then broken by
+comparing `BlockHeaderHashes` of the tips of each chain.
 
-Finally, process any queued children of the newly committed block the same way.
+### `ChainSet` Type
+[chainset-type]: #chainset-type
+
+The `ChainSet` type represents the set of all non-finalized state. It
+consists of a set of non-finalized but verified chains and a set of
+unverified blocks which are waiting for the full context needed to verify
+them to become available.
+
+`ChainState` is defined by the following structure and API:
+
+```rust
+struct ChainSet {
+    chains: BTreeSet<Chain>,
+
+    queued_blocks: BTreeMap<block::Hash, QueuedBlock>,
+    queued_by_parent: BTreeMap<block::Hash, Vec<block::Hash>>,
+    queued_by_height: BTreeMap<block::Height, Vec<block::Hash>>,
+}
+```
+
+#### `pub fn finalize(&mut self) -> Arc<Block>`
+
+Finalize the lowest height block in the non-finalized portion of the best
+chain and updates all side chains to match.
+
+1. Extract the best chain from `self.chains` into `best_chain`
+
+1. Extract the rest of the chains into a `side_chains` temporary variable, so
+   they can be mutated
+
+1. Remove the lowest height block from the best chain with
+   `let block = best_chain.pop_root();`
+
+1. Add `best_chain` back to `self.chains`
+
+1. For each remaining `chain` in `side_chains`
+    - If `chain` starts with `block`, remove `block` and add `chain` back to
+    `self.chains`
+    - Else, drop `chain`
+
+1. for each `height` in `self.queued_by_height` where the height is lower than the
+   new reorg limit
+   - for each `hash` in `self.queued_by_height` at `height`
+     - Remove the key `hash` from `self.queued_blocks` and store the removed `block`
+     - Find and remove `hash` from `self.queued_by_parent` using `block.parent`'s hash
```

Comment:
We also need to remove the corresponding entries from queued_by_height

Reply:
This was supposed to be implied by the `for each` line above, I've added clarification